geomesa-feature/src/main/scala/org/locationtech/geomesa/feature/SimpleFeatureEncoder.scala

```diff
@@ -52,58 +57,114 @@ trait SimpleFeatureEncoder extends HasEncoding {
-            projectedSft: SimpleFeatureType,
-            encoding: String): SimpleFeatureDecoder =
-    SimpleFeatureDecoder(originalSft, projectedSft, FeatureEncoding.withName(encoding))
 }
 
 object SimpleFeatureEncoder {
-  def apply(sft: SimpleFeatureType, encoding: FeatureEncoding): SimpleFeatureEncoder =
+
+  /**
+   * @param sft the simple feature type to be encoded
+   * @param encoding the desired encoding
+   * @param options the desired options
+   * @return a new [[SimpleFeatureEncoder]]
+   */
+  def apply(sft: SimpleFeatureType, encoding: FeatureEncoding, options: Set[EncodingOption] = Set.empty): SimpleFeatureEncoder =
     encoding match {
-      case FeatureEncoding.KRYO => new KryoFeatureEncoder(sft, sft)
-      case FeatureEncoding.AVRO => new AvroFeatureEncoder(sft)
-      case FeatureEncoding.TEXT => new TextFeatureEncoder(sft)
+      case FeatureEncoding.KRYO => new KryoFeatureEncoder(sft, options)
+      case FeatureEncoding.AVRO => new AvroFeatureEncoder(sft, options)
+      case FeatureEncoding.TEXT => new TextFeatureEncoder(sft, options)
     }
-
-  def apply(sft: SimpleFeatureType, encoding: String): SimpleFeatureEncoder =
-    SimpleFeatureEncoder(sft, FeatureEncoding.withName(encoding))
 }
 
-class TextFeatureEncoder(sft: SimpleFeatureType) extends SimpleFeatureEncoder {
-  override def encode(feature:SimpleFeature): Array[Byte] =
-    ThreadSafeDataUtilities.encodeFeature(feature).getBytes
+/**
+ * @param sft the simple feature type to encode
+ * @param options the options to apply when encoding
+ */
+class TextFeatureEncoder(sft: SimpleFeatureType, val options: Set[EncodingOption] = Set.empty)
+  extends SimpleFeatureEncoder {
+
+  private val includeVis = options.contains(EncodingOption.WITH_VISIBILITIES.asInstanceOf[EncodingOption])
+
+  override def encode(feature:SimpleFeature): Array[Byte] = {
+    var encoded = ThreadSafeDataUtilities.encodeFeature(feature)
 
-  override def encoding: FeatureEncoding = FeatureEncoding.TEXT
+    if (includeVis) {
+      val vis = Option(SecurityUtils.getVisibility(feature)).getOrElse("")
+      encoded = TextEncoding.addVisibility(encoded, vis)
+    }
+
+    encoded.getBytes(StandardCharsets.UTF_8)
+  }
+
+  override val encoding: FeatureEncoding = FeatureEncoding.TEXT
 }
 
-class TextFeatureDecoder(sft: SimpleFeatureType) extends SimpleFeatureDecoder {
-  override def decode(bytes: Array[Byte]) =
-    ThreadSafeDataUtilities.createFeature(sft, new String(bytes))
+/**
+ * @param sft the simple feature type to decode
+ * @param options the options what were applied when encoding
+ */
+class TextFeatureDecoder(sft: SimpleFeatureType, val options: Set[EncodingOption] = Set.empty)
+  extends SimpleFeatureDecoder {
+
+  private val expectVis = options.contains(EncodingOption.WITH_VISIBILITIES.asInstanceOf[EncodingOption])
+
+  override def decode(bytes: Array[Byte]): SimpleFeature = {
+    var encoded = new String(bytes, StandardCharsets.UTF_8)
```

Comment:
err, better yet, make this two different methods as vals and pick the right one . . . 

```scala
val defaultDecode: Array[Byte] => SimpleFeature = { bytes =>
   // do default decode
}

val withVizDecode: Array[Byte] => SimpleFeature = { bytes =>
   // do metadata decode
}

override val decode: Array[Byte] => SimpleFeature = if(expectViz) metadataDecode else defaultDecode
```

I think this would be more performant bc it would happen at instantiation time and then the if test would never occur again.